Implement parsing/handling for missing pull request webhook events for BitBucket Server (Stash) driver

raphendyr opened the pull request:
I looked into getting PR events working with the Stash driver to support my teams current workflow. My goal was to address issues mentioned in #116, but I did find few other missing events too.

I open this PR for comments. I have tested these changes with a single user and a single repo in my local BitBucket server, so I'm not yet sure if a PR from a fork work as expected. In addition, I have not written any tests yet.

I'm particularly looking feedback on the parts of the code, which I have added GitHub comments for.

Fixes #116

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/drone/go-scm?pullRequest=130) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/drone/go-scm?pullRequest=130) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/drone/go-scm?pullRequest=130) it.</sub>

raphendyr commented on scm/driver/stash/pr.go:
```diff
@@ -230,6 +230,16 @@ func convertPullRequest(from *pr) *scm.PullRequest {
+		Base: scm.Reference{
+			Name: from.ToRef.DisplayID,
+			Path: from.ToRef.ID,
+			Sha:  from.ToRef.LatestCommit,
```
I think these should be correct. However, I didn't find use for anything else except `.Base.Sha` in Drone server

https://github.com/harness/drone/blob/master/service/hook/parser/parse.go#L275

So, should this addition be omitted from the PR?

raphendyr commented on scm/driver/stash/webhook.go:
```diff
@@ -100,8 +100,24 @@ func (s *webhookService) parsePullRequest(data []byte) (scm.Webhook, error) {
+	case "pr:from_ref_updated":
+		// The request includes field "previousFromHash", which could be compared
+		// to the FromRef.Latestcommit to ensure there is actually a change,
+		// but there is unlikely need for that.
```
Should these comments be added to the git commit instead? They would be available via e.g. blame feature in Github.

raphendyr commented on scm/driver/stash/webhook.go:
```diff
@@ -127,6 +143,16 @@ type pullRequestHook struct {
+		ID              string `json:"id"`              // "refs\/heads\/master"
+		DisplayID       string `json:"displayId"`       // "master"
+		Type            string `json:"type"`            // "BRANCH"
+		LatestCommit    string `json:"latestCommit"`    // "860c4eb4ed0f969b47144234ba13c31c498cca69"
```
Only fields `. PreviousTarget` and `.PreviousTarget.LatestCommit` are currently used. Should I remove everything else from here?

Additionally, if not, should I keep the examples of the data in comments or should I drop those?

bradrydzewski commented:
:wave: hey there, I think in order to provide a review we would need unit tests.  This is largely a data mapping exercise and the unit tests include sample webhooks (with real world data) before and after parsing, which are critical to the review process.

bradrydzewski commented on scm/driver/stash/webhook.go:
```diff
@@ -100,8 +99,24 @@ func (s *webhookService) parsePullRequest(data []byte) (scm.Webhook, error) {
+		// action only when the target reference has changed (name or hash).
+		if src.PullRequest.ToRef.DisplayID == src.PreviousTarget.DisplayID &&
+		src.PullRequest.ToRef.LatestCommit == src.PreviousTarget.LatestCommit {
+			return nil, nil
```
I presume this is checking to differentiate between title / description change vs a code change?  If this is just a title or description change we can set the action to Updated as opposed to returning a nil.

```diff
		if src.PullRequest.ToRef.DisplayID == src.PreviousTarget.DisplayID &&
		src.PullRequest.ToRef.LatestCommit == src.PreviousTarget.LatestCommit {
+			dst.Action = scm.ActionUpdate
		} else {
			dst.Action = scm.ActionSync
		}

```

raphendyr replied:
It is. I'll make that change. I wasn't sure what the `ActionUpdate` was used for.

raphendyr commented:
@bradrydzewski Sure, I was planning to do that next, likely tomorrow. I just wanted to publish what I have done so far for others to read or comment on the things which are not specific to the implementation (i.e., what to include or where to put comments).

raphendyr commented:
I added some test data for the events. I copied the json data from bitbucket server, but I replaced author info to match what was used before.

I noticed that the newer data includes links to different things. So, I think these TODOs might be fixed?

```
webhook.go:// TODO(bradrydzewski) pr hook does not include repository git+http link
webhook.go:// TODO(bradrydzewski) pr hook does not include repository git+ssh link
webhook.go:// TODO(bradrydzewski) pr hook does not include repository html link
```

I can add updated `pr_open.json`, which includes links. I saved a json data from creating a PR from a fork, so I can use that or one without the fork info. Or I can add that as another test, thus yielding one `pr:open` test for in repo PR and test for a PR from a fork.

raphendyr commented:
Also test coverage seems to be bit lower for Stash than others 🤔 

> ok  	github.com/drone/go-scm/scm/driver/stash	0.019s	coverage: 89.3% of statements

d1wilko commented:
hey @raphendyr - I see this is still marked as work in progress, please let us know when you would like us to review :)

raphendyr commented:
> hey @raphendyr - I see this is still marked as work in progress, please let us know when you would like us to review :)

Sorry, my head was stuck waiting possible answers to my questions and without answers, I got stuck there.

tphoney commented:
Hey @raphendyr i have spent some time getting familiar with this PR and the change. Code wise it looks solid and all of the testing is great to have, and made reading through the code much easier.
Going on to your questions 
`Only fields . PreviousTarget and .PreviousTarget.LatestCommit are currently used. Should I remove everything else from here?

Additionally, if not, should I keep the examples of the data in comments or should I drop those?`
please keep the fields.

```
I think these should be correct. However, I didn't find use for anything else except .Base.Sha in Drone server

https://github.com/harness/drone/blob/master/service/hook/parser/parse.go#L275

So, should this addition be omitted from the PR?
```
please add these. it brings it in line with the other transports.

```
Should these comments be added to the git commit instead? They would be available via e.g. blame feature in Github.
```
please leave these in comments in the file, as it aides in reading. 

So leave everything as it is, but could you squash the commits. 

Many thanks for your patience, we have been busy else where. 

TP

raphendyr commented:
@tphoney Thanks for the feedback. As per your request, I'll squash the commits to a single commit.

tphoney commented:
Many thanks @raphendyr for your patience and your excellent code. 🍻
